src/foundation/src/PDFsharp/src/PdfSharp/Pdf.Signatures/DefaultSigner.cs

```diff
@@ -18,6 +19,31 @@ public DefaultSigner(X509Certificate2 Certificate)
             this.Certificate = Certificate;
         }
 
+        public byte[] GetSignedCms(Stream stream, int pdfVersion, byte[] timestampToken)
+        {
+            var range = new byte[stream.Length];
+
+            stream.Position = 0;
+            stream.Read(range, 0, range.Length);
+
+            var contentInfo = new ContentInfo(range);
+            SignedCms signedCms = new SignedCms(contentInfo, true);
+            CmsSigner signer = new CmsSigner(Certificate);
+
+            signer.UnsignedAttributes.Add(new Pkcs9SigningTime());
+
+            if (timestampToken != null)
+            {
+                AsnEncodedData timestampTokenAsnEncodedData = new AsnEncodedData(new Oid("1.2.840.113549.1.9.16.2.14"), timestampToken);
+                signer.UnsignedAttributes.Add(new CryptographicAttributeObject(new Oid("1.2.840.113549.1.9.16.2.14"), new AsnEncodedDataCollection(timestampTokenAsnEncodedData)));
```

Comment:
I read from https://datatracker.ietf.org/doc/html/rfc3126#section-3.12.4 that timestamp must be a signed attribute, but you put it in unsigned attributes. Is it an error or is there a reason?

Reply:
You are right, this need to be fixed.

The document you have referenced is obsoleted by RFC 5126 ([CMS Advanced Electronic Signatures - CAdES](https://datatracker.ietf.org/doc/html/rfc5126)), and [Section 5.11.4](https://datatracker.ietf.org/doc/html/rfc5126#section-5.11.4) states that "The content-time-stamp attribute shall be a signed attribute".

The PAdES standard references CAdES in regards of timestamp attributes (Sections 5.2.2 and 5.3.3 of [this document](https://www.etsi.org/deliver/etsi_ts/102700_102799/10277801/01.01.01_60/ts_10277801v010101p.pdf)).

It says that signatures can optionally include timestamps, which are _recomended_.